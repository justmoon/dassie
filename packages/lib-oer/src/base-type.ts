import { ensureUint8Array } from "./utils/ensure-uint8array"
import { ParseError, SerializeError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"
import { type TagClass, type TagMarker, tagClassMarkerMap } from "./utils/tag"

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyOerType = OerType<any>

export type Infer<TOerType extends AnyOerType> = TOerType extends OerType<
  infer T,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  any
>
  ? T
  : never

export type InferSerialize<TOerType extends AnyOerType> =
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  TOerType extends OerType<any, infer T> ? T : never

export interface ParseOptions {
  allowNoncanonical?: boolean
}

export interface Serializer {
  (context: SerializeContext, offset: number): SerializeError | void
  size: number
}

export abstract class OerType<TParseValue, TSerializeValue = TParseValue> {
  _tag: readonly [tagValue: number, tagClass: TagMarker] | undefined
  _unique = false

  abstract clone(): OerType<TParseValue, TSerializeValue>

  abstract parseWithContext(
    context: ParseContext,
    offset: number
  ): readonly [value: TParseValue, length: number] | ParseError
  abstract serializeWithContext(
    value: TSerializeValue
  ): Serializer | SerializeError

  parse(
    input: Uint8Array,
    offset = 0,
    options: ParseOptions = {}
  ):
    | { success: true; value: TParseValue; length: number }
    | { success: false; error: ParseError } {
    const context: ParseContext = {
      dataView: new DataView(input.buffer, input.byteOffset, input.byteLength),
      uint8Array: ensureUint8Array(input),
      allowNoncanonical: options.allowNoncanonical ?? false,
    }
    const result = this.parseWithContext(context, offset)

    if (result instanceof ParseError) {
      return { success: false, error: result }
    }

    if (!context.allowNoncanonical && result[1] !== input.byteLength - offset) {
      return {
        success: false,
        error: new ParseError(
          "non-canonical encoding - additional bytes present after the expected end of data",
          input,
          result[1]
        ),
      }
    }

    return { success: true, value: result[0], length: result[1] }
  }

  parseOrThrow(
    input: Uint8Array,
    offset = 0,
    options: ParseOptions = {}
  ): TParseValue {
    const result = this.parse(input, offset, options)

    if (result.success) {
      return result.value
    }

    throw result.error
  }

  serialize(
    value: TSerializeValue
  ):
    | { success: true; value: Uint8Array }
    | { success: false; error: SerializeError } {
    const serializer = this.serializeWithContext(value)

    if (serializer instanceof SerializeError) {
      return { success: false, error: serializer }
    }

    const uint8Array = new Uint8Array(serializer.size)

    const result = serializer(
      {
        uint8Array,
        dataView: new DataView(
          uint8Array.buffer,
          uint8Array.byteOffset,
          uint8Array.byteLength
        ),
      },
      0
    )

    return result instanceof SerializeError
      ? { success: false, error: result }
      : { success: true, value: uint8Array }
  }

  serializeOrThrow(value: TSerializeValue): Uint8Array {
    const result = this.serialize(value)

    if (result.success) {
      return result.value
    }

    throw result.error
  }

  tag(tagValue: number, tagClass: Exclude<TagClass, "universal"> = "context") {
    const copy = this.clone()
    copy._tag = [tagValue, tagClassMarkerMap[tagClass]] as const
    return copy
  }

  /**
   * Mark this field as optional.
   *
   * @remarks
   *
   * If the field is not present in the encoding, the value will be `undefined`.
   */
  optional(): OerOptional<TParseValue, TSerializeValue> {
    return new OerOptional(this, undefined)
  }

  /**
   * Provide a default value for this field.
   *
   * @remarks
   *
   * If the field is not present in the encoding, the default value will be used.
   */
  default(value: TParseValue): OerOptional<TParseValue, TSerializeValue> {
    return new OerOptional(this, value)
  }

  /**
   * Mark this field as constant meaning the value must always be the same.
   *
   * @remarks
   *
   * If an encoding contains a different value, parsing will fail.
   */
  constant<TParseValue, TSerializeValue>(
    this: OerType<TParseValue, TSerializeValue>,
    value: TSerializeValue
  ): OerConstant<TParseValue, TSerializeValue> {
    return new OerConstant(this, value)
  }

  /**
   * Marks this type as UNIQUE, which means it will be treated as an identifier field in the context of information objects.
   *
   * @returns A copy of this type with the unique flag set
   */
  unique() {
    const copy = this.clone()
    copy._unique = true
    return copy
  }

  /**
   * Provide a validation function which narrows the type of the parsed value.
   *
   * @remarks
   *
   * This is useful to create custom types which are a subset of the original type.
   *
   * @example
   *
   * ```typescript
   * type PositiveInteger = number & { __positiveInteger: true }
   * const uint8Type = oer.uint8().refind((value): value is PositiveInteger => value > 0))
   * ```
   *
   * @param predicate - A function which returns true if the value is valid
   * @returns A copy of this type with the validation function applied
   */
  refine<TNewType extends TParseValue & TSerializeValue>(
    predicate: (value: TParseValue) => value is TNewType
  ): OerRefined<TNewType, TParseValue, TSerializeValue> {
    return new OerRefined(this, predicate)
  }
}

export class OerOptional<TParseValue, TSerializeValue> extends OerType<
  TParseValue,
  TSerializeValue
> {
  constructor(
    readonly subType: OerType<TParseValue, TSerializeValue>,
    readonly defaultValue: TParseValue | undefined
  ) {
    super()
  }

  clone(): OerOptional<TParseValue, TSerializeValue> {
    return new OerOptional(this.subType, this.defaultValue)
  }

  parseWithContext(
    context: ParseContext,
    offset: number
  ): ParseError | readonly [TParseValue, number] {
    return this.subType.parseWithContext(context, offset)
  }

  serializeWithContext(value: TSerializeValue): SerializeError | Serializer {
    return this.subType.serializeWithContext(value)
  }
}

export class OerConstant<TParseValue, TSerializeValue> extends OerType<
  TParseValue,
  undefined
> {
  private readonly serializedValue: Uint8Array
  private readonly parsedValue: TParseValue

  constructor(
    readonly subType: OerType<TParseValue, TSerializeValue>,
    readonly value: TSerializeValue
  ) {
    super()

    const subtypeSerializeResult = subType.serialize(value)

    if (!subtypeSerializeResult.success) {
      throw new TypeError(
        `Failed to serialize constant OER type with value '${String(value)}'`,
        { cause: subtypeSerializeResult.error }
      )
    }

    this.serializedValue = subtypeSerializeResult.value

    const subtypeParseResult = subType.parse(this.serializedValue)

    if (!subtypeParseResult.success) {
      throw new TypeError(
        `Failed to parse constant OER type with value '${String(value)}'`,
        { cause: subtypeParseResult.error }
      )
    }

    this.parsedValue = subtypeParseResult.value
  }

  clone() {
    return new OerConstant(this.subType, this.value)
  }

  parseWithContext(
    context: ParseContext,
    offset: number
  ): ParseError | readonly [TParseValue, number] {
    if (context.uint8Array.length < offset + this.serializedValue.length) {
      return new ParseError(
        `unable to read constant value - end of buffer`,
        context.uint8Array,
        context.uint8Array.length
      )
    }

    for (let index = 0; index < this.serializedValue.length; index++) {
      if (context.uint8Array[offset + index] !== this.serializedValue[index]) {
        return new ParseError(
          `expected constant value did not match data while parsing`,
          context.uint8Array,
          offset + index
        )
      }
    }

    return [this.parsedValue, this.serializedValue.length]
  }

  serializeWithContext(): Serializer {
    // eslint-disable-next-line unicorn/consistent-function-scoping
    const serializer = (context: SerializeContext, offset: number) => {
      context.uint8Array.set(this.serializedValue, offset)
    }
    serializer.size = this.serializedValue.length
    return serializer
  }

  override serialize() {
    return super.serialize(undefined)
  }
}

export class OerRefined<
  TNarrowedType extends TParseType & TSerializeValue,
  TParseType,
  TSerializeValue
> extends OerType<TNarrowedType, TNarrowedType> {
  constructor(
    readonly subType: OerType<TParseType, TSerializeValue>,
    readonly predicate: (value: TParseType) => value is TNarrowedType
  ) {
    super()
  }

  clone(): OerRefined<TNarrowedType, TParseType, TSerializeValue> {
    return new OerRefined(this.subType, this.predicate)
  }

  parseWithContext(
    context: ParseContext,
    offset: number
  ): ParseError | readonly [TNarrowedType, number] {
    const result = this.subType.parseWithContext(context, offset)

    if (result instanceof ParseError) {
      return result
    }

    const [value, size] = result

    if (!this.predicate(value)) {
      return new ParseError(
        `refinement predicate failed`,
        context.uint8Array,
        offset
      )
    }

    return [value, size]
  }

  serializeWithContext(value: TNarrowedType): SerializeError | Serializer {
    return this.subType.serializeWithContext(value)
  }
}
