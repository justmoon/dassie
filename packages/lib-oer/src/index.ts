export { OerType, OerOptional, OerConstant, OerRefined } from "./base-type"
export type {
  AnyOerType,
  ParseOptions,
  Serializer,
  Infer,
  InferSerialize,
} from "./base-type"

export * from "./bitstring"
export * from "./boolean"
export * from "./captured"
export * from "./choice"
export * from "./empty"
export * from "./enumerated"
export * from "./integer-as-bigint"
export * from "./integer-as-number"
export * from "./object-identifier"
export * from "./octet-string"
export * from "./sequence-of"
export * from "./sequence"
export * from "./string"

export * from "./information-object/class"
export * from "./information-object/object-set"

export * from "./utils/alphabet"

export { uint8ArrayToHex, hexToUint8Array } from "./utils/hex"
export { ParseFailure, SerializeFailure } from "./utils/failures"
