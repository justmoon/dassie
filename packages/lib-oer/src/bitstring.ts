import type { ReadonlyTuple } from "type-fest"

import { OerType } from "./base-type"
import { ParseError, SerializeError } from "./utils/errors"
import {
  parseLengthPrefix,
  predictLengthPrefixLength,
  serializeLengthPrefix,
} from "./utils/length-prefix"
import type { BaseContext, ParseContext, SerializeContext } from "./utils/parse"

type InferBitstringValue<T extends number> = ReadonlyTuple<boolean, T>

interface BitstringOptions {
  variableLength?: boolean
}

export const parseBitstring = (
  context: ParseContext,
  offset: number,
  size: number,
) => {
  const byteLength = Math.ceil(size / 8)
  const output = Array.from({ length: size })
  for (let byteIndex = 0; byteIndex < byteLength; byteIndex++) {
    const byte = context.uint8Array[offset + byteIndex]!

    for (let bitIndex = 0; bitIndex < 8; bitIndex++) {
      if (byteIndex * 8 + bitIndex >= size) {
        break
      }
      const bit = (byte >> (7 - bitIndex)) & 1
      output[byteIndex * 8 + bitIndex] = bit === 1
    }
  }
  return output
}

export const serializeBitstring = (
  context: BaseContext,
  offset: number,
  input: InferBitstringValue<number>,
) => {
  const { uint8Array } = context
  const inputAsArray = input as boolean[]
  const byteLength = Math.ceil(inputAsArray.length / 8)

  for (let byteIndex = 0; byteIndex < byteLength; byteIndex++) {
    let byte = 0
    for (let bitIndex = 0; bitIndex < 8; bitIndex++) {
      const index = byteIndex * 8 + bitIndex
      if (index >= inputAsArray.length) break
      byte |= inputAsArray[index] ? 1 << (7 - bitIndex) : 0
    }
    uint8Array[offset + byteIndex] = byte
  }
}

export class OerVariableBitstring<
  TBitDefinition extends number,
> extends OerType<InferBitstringValue<TBitDefinition>> {
  constructor(readonly bits: TBitDefinition) {
    super()
  }

  clone() {
    return new OerVariableBitstring(this.bits)
  }

  parseWithContext(context: ParseContext, offset: number) {
    const { uint8Array } = context
    const result = parseLengthPrefix(context, offset)

    if (result instanceof ParseError) {
      return result
    }

    const [length, lengthOfLength] = result

    const unusedBits = uint8Array[offset + lengthOfLength]!

    if (unusedBits > 7) {
      return new ParseError(
        "unable to read bitstring - unused bits greater than 7",
        uint8Array,
        offset + 1,
      )
    }

    const output = parseBitstring(
      context,
      offset + lengthOfLength + 1,
      (length - 1) * 8 - unusedBits,
    )

    return [
      output as InferBitstringValue<TBitDefinition>,
      lengthOfLength + length,
    ] as const
  }

  serializeWithContext(input: ReadonlyTuple<boolean, TBitDefinition>) {
    const byteLength = Math.ceil(this.bits / 8)
    const unusedBits = 8 - (this.bits % 8)
    const length = byteLength + 1
    const lengthOfLength = predictLengthPrefixLength(length)
    if (lengthOfLength instanceof SerializeError) {
      return lengthOfLength
    }

    const serializer = (context: SerializeContext, offset: number) => {
      const { uint8Array } = context
      const lengthOfLength = serializeLengthPrefix(length, uint8Array, offset)

      if (lengthOfLength instanceof SerializeError) {
        return lengthOfLength
      }

      uint8Array[offset + lengthOfLength] = unusedBits
      serializeBitstring(context, offset + lengthOfLength + 1, input)
      return
    }
    serializer.size = lengthOfLength + length
    return serializer
  }
}

export class OerFixedBitstring<TBitDefinition extends number> extends OerType<
  InferBitstringValue<TBitDefinition>
> {
  constructor(readonly bits: TBitDefinition) {
    super()
  }

  clone() {
    return new OerFixedBitstring(this.bits)
  }

  parseWithContext(context: ParseContext, offset: number) {
    const byteLength = Math.ceil(this.bits / 8)

    if (context.uint8Array.length - offset < byteLength) {
      return new ParseError(
        "unable to read bitstring value - end of buffer",
        context.uint8Array,
        offset,
      )
    }

    const output = parseBitstring(context, offset, this.bits)

    return [output as InferBitstringValue<TBitDefinition>, byteLength] as const
  }

  serializeWithContext(input: InferBitstringValue<TBitDefinition>) {
    const inputAsArray = input as boolean[]
    if (inputAsArray.length !== this.bits) {
      return new SerializeError(
        `unable to serialize bitstring - expected ${this.bits} bits, got ${inputAsArray.length}`,
      )
    }

    const serializer = (context: SerializeContext, offset: number) => {
      serializeBitstring(context, offset, input)
    }
    serializer.size = Math.ceil(this.bits / 8)

    return serializer
  }
}

export const bitstring = <TBitDefinition extends number>(
  bits: TBitDefinition,
  { variableLength = false }: BitstringOptions = {},
) => {
  return variableLength
    ? new OerVariableBitstring<TBitDefinition>(bits)
    : new OerFixedBitstring<TBitDefinition>(bits)
}
