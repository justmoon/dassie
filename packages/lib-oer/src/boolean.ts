import { OerType } from "./base-type"
import { ParseError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"

export class OerBoolean extends OerType<boolean> {
  clone() {
    return new OerBoolean()
  }

  parseWithContext(
    { uint8Array, allowNoncanonical }: ParseContext,
    offset: number,
  ) {
    const value = uint8Array[offset]
    if (value === undefined) {
      return new ParseError(
        "unable to read boolean value - end of buffer",
        uint8Array,
        offset,
      )
    }
    if (!allowNoncanonical && value !== 0 && value !== 0xff) {
      return new ParseError(
        "unable to read boolean value - non-canonical",
        uint8Array,
        offset,
      )
    }
    return [!!value, 1] as const
  }
  serializeWithContext(input: boolean) {
    const serializer = ({ uint8Array }: SerializeContext, offset: number) => {
      uint8Array[offset] = input ? 0xff : 0
    }
    serializer.size = 1
    return serializer
  }
}

export const boolean = () => {
  return new OerBoolean()
}
