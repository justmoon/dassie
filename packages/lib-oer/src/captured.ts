import { OerType } from "./base-type"
import { ParseError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"

export class OerCaptured<TParseValue, TSerializeValue> extends OerType<
  { value: TParseValue; bytes: Uint8Array },
  { value: TSerializeValue } | { bytes: Uint8Array }
> {
  constructor(readonly subType: OerType<TParseValue, TSerializeValue>) {
    super()
  }

  clone() {
    return new OerCaptured(this.subType)
  }

  parseWithContext(context: ParseContext, offset: number) {
    const subTypeResult = this.subType.parseWithContext(context, offset)

    if (subTypeResult instanceof ParseError) {
      return subTypeResult
    }

    const [subTypeValue, subTypeLength] = subTypeResult

    return [
      {
        value: subTypeValue,
        bytes: new Uint8Array(
          context.uint8Array.buffer,
          context.uint8Array.byteOffset + offset,
          subTypeLength,
        ),
      },
      subTypeLength,
    ] as const
  }

  serializeWithContext(
    input: { value: TSerializeValue } | { bytes: Uint8Array },
  ) {
    if ("value" in input) {
      return this.subType.serializeWithContext(input.value)
    } else {
      const serializer = (context: SerializeContext, offset: number) => {
        context.uint8Array.set(input.bytes, offset)
      }
      serializer.size = input.bytes.length
      return serializer
    }
  }
}

export const captured = <TParseValue, TSerializeValue>(
  subType: OerType<TParseValue, TSerializeValue>,
) => {
  return new OerCaptured(subType)
}
