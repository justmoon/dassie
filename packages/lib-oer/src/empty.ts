import { OerType } from "./base-type"

const serializer = () => {
  // no-op
}
serializer.size = 0

export class OerEmpty extends OerType<void> {
  clone() {
    return new OerEmpty()
  }

  parseWithContext() {
    return [undefined, 0] as const
  }
  serializeWithContext() {
    return serializer
  }
}

export const empty = () => {
  return new OerEmpty()
}
