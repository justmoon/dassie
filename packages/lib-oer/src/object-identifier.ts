import { assertDefined, isFailure } from "@dassie/lib-type-utils"

import { OerType } from "./base-type"
import {
  parseBase128,
  predictBase128Length,
  serializeBase128,
} from "./utils/base-128"
import { ParseFailure, SerializeFailure } from "./utils/failures"
import {
  parseLengthPrefix,
  predictLengthPrefixLength,
  serializeLengthPrefix,
} from "./utils/length-prefix"
import type { ParseContext, SerializeContext } from "./utils/parse"

export class OerObjectIdentifier extends OerType<string> {
  clone() {
    return new OerObjectIdentifier()
  }

  parseWithContext(context: ParseContext, offset: number) {
    const { uint8Array } = context

    const result = parseLengthPrefix(context, offset)
    if (isFailure(result)) return result

    const [length, lengthOfLength] = result

    if (length === 0) {
      return new ParseFailure(
        "object identifier of length zero is invalid",
        uint8Array,
        offset,
      )
    }

    const subidentifiers: bigint[] = []

    let nextOffset = offset + lengthOfLength //?
    while (nextOffset < offset + lengthOfLength + length) {
      const subidentifier = parseBase128(
        context,
        nextOffset,
        offset + lengthOfLength + length,
      )
      if (isFailure(subidentifier)) return subidentifier

      subidentifiers.push(subidentifier[0])
      nextOffset += subidentifier[1]
    }

    assertDefined(subidentifiers[0])

    const firstTwo =
      subidentifiers[0] >= 80
        ? ([2, subidentifiers[0] - 80n] as const)
        : ([subidentifiers[0] / 40n, subidentifiers[0] % 40n] as const)

    return [
      [...firstTwo, ...subidentifiers.slice(1)].join("."),
      nextOffset - offset,
    ] as const
  }

  serializeWithContext(input: string) {
    const objectIdentifierComponents = input.split(".").map(BigInt)

    if (objectIdentifierComponents.some((x) => x < 0n)) {
      return new SerializeFailure(
        "object identifier components must be non-negative",
      )
    }

    if (objectIdentifierComponents.length < 2) {
      return new SerializeFailure(
        "object identifier must have at least two components",
      )
    }

    if (
      objectIdentifierComponents[0]! > 2 ||
      objectIdentifierComponents[0]! < 0
    ) {
      return new SerializeFailure(
        "object identifier first component must be in the range of 0..2",
      )
    }

    if (
      objectIdentifierComponents[0]! < 2 &&
      objectIdentifierComponents[1]! >= 40
    ) {
      return new SerializeFailure(
        "object identifier second component must be in the range of 0..39 when first component is 0 or 1",
      )
    }

    const firstSubidentifier =
      objectIdentifierComponents[0]! * 40n + objectIdentifierComponents[1]!

    const subidentifiers = [
      firstSubidentifier,
      ...objectIdentifierComponents.slice(2),
    ]

    let length = 0

    for (const subidentifier of subidentifiers) {
      const subidentifierLength = predictBase128Length(subidentifier)
      if (isFailure(subidentifierLength)) return subidentifierLength

      length += subidentifierLength
    }

    const lengthOfLengthPrefix = predictLengthPrefixLength(length)
    if (isFailure(lengthOfLengthPrefix)) return lengthOfLengthPrefix

    const serializer = ({ uint8Array }: SerializeContext, offset: number) => {
      const lengthPrefixSerializeResult = serializeLengthPrefix(
        length,
        uint8Array,
        offset,
      )

      if (isFailure(lengthPrefixSerializeResult))
        return lengthPrefixSerializeResult

      let currentOffset = (offset = lengthOfLengthPrefix)
      for (const subidentifier of subidentifiers) {
        const subidentifierSerializeResult = serializeBase128(
          subidentifier,
          uint8Array,
          currentOffset,
        )

        if (isFailure(subidentifierSerializeResult))
          return subidentifierSerializeResult

        currentOffset += subidentifierSerializeResult
      }
      return
    }
    serializer.size = lengthOfLengthPrefix + length
    return serializer
  }
}

export const objectIdentifier = () => {
  return new OerObjectIdentifier()
}
