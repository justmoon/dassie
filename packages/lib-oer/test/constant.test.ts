import { describe, test } from "vitest"

import { boolean, hexToUint8Array, sequence, uint8Number } from "../src"
import { objectIdentifier } from "../src/object-identifier"
import { constantTests } from "./utils/full-samples"
import { parsedOk, serializedOk } from "./utils/result"

describe("constant", () => {
  describe.each(constantTests)("of %s", (_, baseType, value, hex) => {
    const schema = baseType.constant(value)

    test("should return an object", ({ expect }) => {
      expect(schema).toBeTypeOf("object")
    })

    test("should parse the correct value", ({ expect }) => {
      const result = schema.parse(hexToUint8Array(hex))
      expect(result).toEqual(parsedOk(hex.length / 2, value))
    })

    test("should serialize a valid value", ({ expect }) => {
      const value = schema.serialize()
      expect(value).toEqual(serializedOk(hex))
    })

    test("should fail to parse a value extended by one byte", ({ expect }) => {
      const result = schema.parse(hexToUint8Array(`${hex}00`))
      expect(result).toMatchSnapshot()
    })

    if (hex.length >= 2) {
      test("should fail to parse a value truncated by one byte", ({
        expect,
      }) => {
        const result = schema.parse(hexToUint8Array(hex.slice(0, -2)))
        expect(result).toMatchSnapshot()
      })

      test("should fail to parse a value with a different encoding than expected", ({
        expect,
      }) => {
        const result = schema.parse(
          hexToUint8Array(
            ((Number.parseInt(hex.slice(0, 2), 16) + 1) % 256)
              .toString(16)
              .padStart(2, "0") + hex.slice(2),
          ),
        )
        expect(result).toMatchSnapshot()
      })
    }
  })

  describe("inside of a sequence", () => {
    const schema = sequence({
      a: uint8Number(),
      b: objectIdentifier().constant("2.999"),
      c: boolean(),
    })

    test("should return an object", ({ expect }) => {
      expect(schema).toBeTypeOf("object")
    })

    test("should parse a valid value", ({ expect }) => {
      const value = schema.parse(hexToUint8Array("01 02 88 37 FF"))
      expect(value).toEqual(parsedOk(5, { a: 1, b: "2.999", c: true }))
    })

    test("should serialize a valid value", ({ expect }) => {
      // Note that we do not need to provide the constant value during serialization.
      const value = schema.serialize({ a: 1, c: true })
      expect(value).toEqual(serializedOk("01 02 88 37 FF"))
    })
  })
})
