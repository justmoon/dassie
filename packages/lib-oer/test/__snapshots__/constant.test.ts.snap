// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`constant > of bitstring > should fail to parse a value extended by one byte 1`] = `
[ParseFailure(offset 1): non-canonical encoding - additional bytes present after the expected end of data

    a0 00  
       ^^]
`;

exports[`constant > of bitstring > should fail to parse a value truncated by one byte 1`] = `
[ParseFailure(offset 0): unable to read constant value - end of buffer

      
    ^^]
`;

exports[`constant > of bitstring > should fail to parse a value with a different encoding than expected 1`] = `
[ParseFailure(offset 0): expected constant value did not match data while parsing

    a1  
    ^^]
`;

exports[`constant > of boolean > should fail to parse a value extended by one byte 1`] = `
[ParseFailure(offset 1): non-canonical encoding - additional bytes present after the expected end of data

    ff 00  
       ^^]
`;

exports[`constant > of boolean > should fail to parse a value truncated by one byte 1`] = `
[ParseFailure(offset 0): unable to read constant value - end of buffer

      
    ^^]
`;

exports[`constant > of boolean > should fail to parse a value with a different encoding than expected 1`] = `
[ParseFailure(offset 0): expected constant value did not match data while parsing

    00  
    ^^]
`;

exports[`constant > of empty fixed size octet string > should fail to parse a value extended by one byte 1`] = `
[ParseFailure(offset 0): non-canonical encoding - additional bytes present after the expected end of data

    00  
    ^^]
`;

exports[`constant > of fixed size octet string > should fail to parse a value extended by one byte 1`] = `
[ParseFailure(offset 4): non-canonical encoding - additional bytes present after the expected end of data

    12 34 56 78 00  
                ^^]
`;

exports[`constant > of fixed size octet string > should fail to parse a value truncated by one byte 1`] = `
[ParseFailure(offset 3): unable to read constant value - end of buffer

    12 34 56  
             ^^]
`;

exports[`constant > of fixed size octet string > should fail to parse a value with a different encoding than expected 1`] = `
[ParseFailure(offset 0): expected constant value did not match data while parsing

    13 34 56 78  
    ^^]
`;

exports[`constant > of sequence > should fail to parse a value extended by one byte 1`] = `
[ParseFailure(offset 3): non-canonical encoding - additional bytes present after the expected end of data

    0f 00 de 00  
             ^^]
`;

exports[`constant > of sequence > should fail to parse a value truncated by one byte 1`] = `
[ParseFailure(offset 2): unable to read constant value - end of buffer

    0f 00  
          ^^]
`;

exports[`constant > of sequence > should fail to parse a value with a different encoding than expected 1`] = `
[ParseFailure(offset 0): expected constant value did not match data while parsing

    10 00 de  
    ^^]
`;

exports[`constant > of sequenceOf > should fail to parse a value extended by one byte 1`] = `
[ParseFailure(offset 10): non-canonical encoding - additional bytes present after the expected end of data

    01 04 f1 00 f2 00 f3 00 f4 00 00  
                                  ^^]
`;

exports[`constant > of sequenceOf > should fail to parse a value truncated by one byte 1`] = `
[ParseFailure(offset 9): unable to read constant value - end of buffer

    01 04 f1 00 f2 00 f3 00 f4  
                               ^^]
`;

exports[`constant > of sequenceOf > should fail to parse a value with a different encoding than expected 1`] = `
[ParseFailure(offset 0): expected constant value did not match data while parsing

    02 04 f1 00 f2 00 f3 00 f4 00  
    ^^]
`;

exports[`constant > of uint8 > should fail to parse a value extended by one byte 1`] = `
[ParseFailure(offset 1): non-canonical encoding - additional bytes present after the expected end of data

    0f 00  
       ^^]
`;

exports[`constant > of uint8 > should fail to parse a value truncated by one byte 1`] = `
[ParseFailure(offset 0): unable to read constant value - end of buffer

      
    ^^]
`;

exports[`constant > of uint8 > should fail to parse a value with a different encoding than expected 1`] = `
[ParseFailure(offset 0): expected constant value did not match data while parsing

    10  
    ^^]
`;

exports[`constant > of uint64Bigint > should fail to parse a value extended by one byte 1`] = `
[ParseFailure(offset 8): non-canonical encoding - additional bytes present after the expected end of data

    ff ff ff ff ff ff ff ff 00  
                            ^^]
`;

exports[`constant > of uint64Bigint > should fail to parse a value truncated by one byte 1`] = `
[ParseFailure(offset 7): unable to read constant value - end of buffer

    ff ff ff ff ff ff ff  
                         ^^]
`;

exports[`constant > of uint64Bigint > should fail to parse a value with a different encoding than expected 1`] = `
[ParseFailure(offset 0): expected constant value did not match data while parsing

    00 ff ff ff ff ff ff ff  
    ^^]
`;

exports[`constant > of variable length octet string > should fail to parse a value extended by one byte 1`] = `
[ParseFailure(offset 5): non-canonical encoding - additional bytes present after the expected end of data

    04 12 34 56 78 00  
                   ^^]
`;

exports[`constant > of variable length octet string > should fail to parse a value truncated by one byte 1`] = `
[ParseFailure(offset 4): unable to read constant value - end of buffer

    04 12 34 56  
                ^^]
`;

exports[`constant > of variable length octet string > should fail to parse a value with a different encoding than expected 1`] = `
[ParseFailure(offset 0): expected constant value did not match data while parsing

    05 12 34 56 78  
    ^^]
`;
